pkg/updatestatus/nodeinformer.go

```diff
@@ -0,0 +1,390 @@
+		return true
+	}
+	return false
+}
+
+func isNodeDraining(node *corev1.Node, isUpdating bool) bool {
+	desiredDrain := node.Annotations[mco.DesiredDrainerAnnotationKey]
+	appliedDrain := node.Annotations[mco.LastAppliedDrainerAnnotationKey]
+
+	if appliedDrain == "" || desiredDrain == "" {
+		return false
+	}
+
+	if desiredDrain != appliedDrain {
+		desiredVerb := strings.Split(desiredDrain, "-")[0]
+		if desiredVerb == mco.DrainerStateDrain {
+			return true
+		}
+	}
+
+	// Node is supposed to be updating but MCD hasn't had the time to update
+	// its state from original `Done` to `Working` and start the drain process.
+	// Default to drain process so that we don't report completed.
+	mcdState := node.Annotations[mco.MachineConfigDaemonStateAnnotationKey]
+	return isUpdating && mcdState == mco.MachineConfigDaemonStateDone
+}
+
+func determineConditions(pool *machineconfigv1.MachineConfigPool, node *corev1.Node, isUpdating, isUpdated, isUnavailable, isDegraded bool, lns *mco.LayeredNodeState, now metav1.Time) ([]metav1.Condition, string, *metav1.Duration) {
+	var estimate *metav1.Duration
+	var message string
+
+	updating := metav1.Condition{
+		Type:               string(NodeStatusInsightUpdating),
+		Status:             metav1.ConditionUnknown,
+		Reason:             string(NodeCannotDetermine),
+		LastTransitionTime: now,
+	}
+	available := metav1.Condition{
+		Type:               string(NodeStatusInsightAvailable),
+		Status:             metav1.ConditionTrue,
+		LastTransitionTime: now,
+	}
+	degraded := metav1.Condition{
+		Type:               string(NodeStatusInsightDegraded),
+		Status:             metav1.ConditionFalse,
+		LastTransitionTime: now,
+	}
+
+	if isUpdating && isNodeDraining(node, isUpdating) {
+		estimate = toPointer(10 * time.Minute)
+		updating.Status = metav1.ConditionTrue
+		updating.Reason = string(NodeDraining)
+	} else if isUpdating {
+		state := node.Annotations[mco.MachineConfigDaemonStateAnnotationKey]
+		switch state {
+		case mco.MachineConfigDaemonStateRebooting:
+			estimate = toPointer(10 * time.Minute)
+			updating.Status = metav1.ConditionTrue
+			updating.Reason = string(NodeRebooting)
+		case mco.MachineConfigDaemonStateDone:
+			estimate = toPointer(time.Duration(0))
+			updating.Status = metav1.ConditionFalse
+			updating.Reason = string(NodeCompleted)
+		default:
+			estimate = toPointer(10 * time.Minute)
+			updating.Status = metav1.ConditionTrue
+			updating.Reason = string(NodeUpdating)
+		}
+
+	} else if isUpdated {
+		estimate = toPointer(time.Duration(0))
+		updating.Status = metav1.ConditionFalse
+		updating.Reason = string(NodeCompleted)
+	} else if pool.Spec.Paused {
+		estimate = toPointer(time.Duration(0))
+		updating.Status = metav1.ConditionFalse
+		updating.Reason = string(NodePaused)
+	} else {
+		updating.Status = metav1.ConditionFalse
+		updating.Reason = string(NodeUpdatePending)
+	}
+	message = updating.Message
```

Comment:
...and because condition Message is required, but status insight can be empty (for boring cases), we will need to only put into insight message the information that is actually interesting.